src/values/AbstractObjectValue.js

```diff
@@ -668,26 +720,66 @@ export default class AbstractObjectValue extends AbstractValue {
 
   $SetPartial(P: AbstractValue | PropertyKeyValue, V: Value, Receiver: Value): boolean {
     if (!(P instanceof AbstractValue)) return this.$Set(P, V, Receiver);
-    invariant(this === Receiver, "TODO #1021");
     if (this.values.isTop()) {
-      AbstractValue.reportIntrospectionError(this);
+      if (this.$Realm.isInPureScope()) {
```

Comment:
This seems to be a duplicate of the code in line 632. You can avoid the duplication by moving line 722 below this if.

Reply:
Not all callsites go through $SetPartial though. I guess ideally they should tho, right?

Reply:
Meanwhile you could call $SetPartial from $Set in this particular case.